server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -172,9 +175,39 @@ protected Set<String> modifiedKeys(Request request) {
         return Collections.emptySet();
     }
 
+    // package private for testing
+    void validateForImmutableState(Request request, ClusterState state) {
+        Optional<String> handlerName = immutableStateHandlerName();
+        assert handlerName.isPresent();
+
+        Set<String> modified = modifiedKeys(request);
+        List<String> errors = new ArrayList<>();
+
+        for (ImmutableStateMetadata metadata : state.metadata().immutableStateMetadata().values()) {
+            Set<String> conflicts = metadata.conflicts(handlerName.get(), modified);
+            if (conflicts.isEmpty() == false) {
+                errors.add(format("[%s] set as read-only by [%s]", String.join(",", conflicts), metadata.namespace()));
+            }
+        }
+
+        if (errors.isEmpty() == false) {
+            throw new IllegalArgumentException(
+                format("Failed to process request [%s] with errors: %s", request, String.join(System.lineSeparator(), errors))
```

Comment:
I don't know if we have multiline error messages in other exceptions. Will this IAE propgate through a user request (that would return a 400)? eg a user tries to add an ILM policy with an already reserved name, then they get this exception. If we want to go with newlines, I think we need to make sure this gets serialized correctly (ie the newline needs to be escaped in the json response).

Also, I'm not sure System.lineSeparator is necessary? At least for json only `\n` would be expected.

Reply:
OK, sounds good. I'll wrap the errors as an array with [] and commas. That should solve the windows line separator test issue too.

Reply:
> Will this IAE propgate through a user request (that would return a 400)?

Yes it does go all the way back to the user that issued the REST request.